docs/dropout.md

```diff
@@ -1 +1,402 @@
+def fig_to_html(fig: Figure) -> str:  # markdown-exec: hide
+    buffer = StringIO()  # markdown-exec: hide
+    fig.savefig(buffer, format="svg")  # markdown-exec: hide
+    return buffer.getvalue()  # markdown-exec: hide
+print(fig_to_html(plt.gcf())) # markdown-exec: hide
+```
+
+Since our QNN can only output values between -1 and 1 we need to scale the data to be between these values.
+
+```python exec="on" source="material-block" session="dropout"
+class MinMaxScaler:
+    """A class which scales data to be within a chosen range"""
+
+    def __init__(self, feature_range: tuple =(0, 1)):
+        self.feature_range = feature_range
+        self.min = None
+        self.scale = None
+
+    def fit(self, X):
+        self.min = X.min(axis=0)
+        self.scale = X.max(axis=0) - self.min
+        self.scale[self.scale == 0] = 1  # Avoid division by zero for constant features
+
+    def transform(self, X):
+        X_scaled = (X - self.min) / self.scale
+        X_scaled = (
+            X_scaled * (self.feature_range[1] - self.feature_range[0]) + self.feature_range[0]
+        )
+        return X_scaled
+
+    def fit_transform(self, X):
+        self.fit(X)
+        return self.transform(X)
+
+    def inverse_transform(self, X_scaled):
+        X = (X_scaled - self.feature_range[0]) / (self.feature_range[1] - self.feature_range[0])
+        X = X * self.scale + self.min
+        return X
+
+scaler = MinMaxScaler(feature_range=(-1, 1))
+y_train = scaler.fit_transform(y_train)
+y_test = scaler.transform(y_test)
+```
+
+Now we need to construct our QNNs, they are comprised of feature maps and an ansatz. The ansatz contains CNOT gates with nearest neighbour entanglement after every rotation gate. The feature map contains two rotation gates RY and RZ, taking as rotation angles $\arcsin(x)$ and $\arccos(x^2)$ respectively.
+
+```python exec="on" source="material-block" session="dropout"
+
+def hea_ansatz(n_qubits, layer):
+    """creates an ansatz which performs RX, RZ, RX rotations on each qubit,
+    which nearest neighbour CNOT gates interpersed between each rotational gate."""
+    ops = []
+    for i in range(n_qubits):
+        ops.append(pyq.RX(i, param_name=f"theta_{i}{layer}{0}"))
+
+    for j in range(n_qubits - 1):
+        ops.append(pyq.CNOT(control=j, target=j + 1))
+
+    for i in range(n_qubits):
+        ops.append(pyq.RZ(i, param_name=f"theta_{i}{layer}{1}"))
+
+    for j in range(n_qubits - 1):
+        ops.append(pyq.CNOT(control=j, target=j + 1))
+
+    for i in range(n_qubits):
+        ops.append(pyq.RX(i, param_name=f"theta_{i}{layer}{2}"))
+
+    for j in range(n_qubits - 1):
+        ops.append(pyq.CNOT(control=j, target=j + 1))
+    return ops
+
+# the two feature maps we will be using
+def fm1(n_qubits):
+    return [pyq.RY(i, "x1") for i in range(n_qubits)]
+
+
+def fm2(n_qubits):
+    return [pyq.RZ(i, "x2") for i in range(n_qubits)]
+
+# The class which constructs QNNs
+class QuantumModelBase(torch.nn.Module):
```

Comment:
I'm a bit surprised there is no datastructure you could use here instead of reconstructing one.

Reply:
No, there isn't. So far there wasn't really a need for one. 

It seems to me like the usage here is just to follow the typical torch neural network construction, which in a way is a more "torchy" use of PyQ than what Qadence actually does. 

For now, if it's just for convenience of the tutorial that's fine. We can then see if a general version of this deserves a place in PyQ.